src/expr/src/expr/expr_vnode.rs

```diff
@@ -0,0 +1,181 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+
+use risingwave_common::array::{
+    ArrayBuilder, ArrayImpl, ArrayRef, DataChunk, I32ArrayBuilder, Row,
+};
+use risingwave_common::types::{DataType, Datum};
+use risingwave_common::util::hash_util::CRC32FastBuilder;
+use risingwave_pb::expr::expr_node::{RexNode, Type};
+use risingwave_pb::expr::ExprNode;
+
+use super::{build_from_prost as expr_build_from_prost, BoxedExpression, Expression};
+use crate::{bail, ensure, ExprError, Result};
+
+#[derive(Debug)]
+pub struct VnodeExpression {
+    dist_key_exprs: Vec<BoxedExpression>,
+}
+
+impl VnodeExpression {
+    pub fn new(dist_key_exprs: Vec<BoxedExpression>) -> Self {
+        VnodeExpression { dist_key_exprs }
+    }
+}
+
+impl<'a> TryFrom<&'a ExprNode> for VnodeExpression {
+    type Error = ExprError;
+
+    fn try_from(prost: &'a ExprNode) -> Result<Self> {
+        ensure!(prost.get_expr_type().unwrap() == Type::Vnode);
+        ensure!(DataType::from(prost.get_return_type().unwrap()) == DataType::Int32);
+
+        let RexNode::FuncCall(func_call_node) = prost.get_rex_node().unwrap() else {
+            bail!("Expected RexNode::FuncCall");
+        };
+
+        let dist_key_exprs = func_call_node
+            .children
+            .iter()
+            .map(expr_build_from_prost)
+            .collect::<Result<Vec<_>>>()?;
+
+        Ok(VnodeExpression::new(dist_key_exprs))
+    }
+}
+
+impl Expression for VnodeExpression {
+    fn return_type(&self) -> DataType {
+        DataType::Int32
```

Comment:
Int16 should be enough?
```suggestion
        DataType::Int16
```